src/llvm-late-gc-lowering.cpp

```diff
@@ -1889,6 +1902,189 @@ void LateLowerGCFrame::PlaceGCFrameStores(State &S, unsigned MinColorRoot,
     }
 }
 
+template<typename T1, typename T2, typename DT>
+static void HandleOnePushPopEdge(T1 &edges, T2 &ins_vec, DT *dt)
+{
+    auto edge = edges.pop_back_val();
+    // First check if we can avoid splitting the edge
+    bool all_edge_in_set = true;
+    for (auto pred_bb: predecessors(edge.second)) {
+        if (edge.first == pred_bb)
+            continue;
+        if (edges.count(std::make_pair(pred_bb, edge.second)) == 0) {
+            all_edge_in_set = false;
+            break;
+        }
+    }
+    if (all_edge_in_set) {
+        for (auto pred_bb: predecessors(edge.second)) {
+            if (edge.first == pred_bb)
+                continue;
+            edges.remove(std::make_pair(pred_bb, edge.second));
+        }
+        ins_vec.push_back(edge.second->getFirstNonPHI());
+        return;
+    }
+    all_edge_in_set = true;
+    for (auto succ_bb: successors(edge.first)) {
+        if (edge.second == succ_bb)
+            continue;
+        if (edges.count(std::make_pair(edge.first, succ_bb)) == 0) {
+            all_edge_in_set = false;
+            break;
+        }
+    }
+    if (all_edge_in_set) {
+        for (auto succ_bb: successors(edge.first)) {
+            if (edge.second == succ_bb)
+                continue;
+            edges.remove(std::make_pair(edge.first, succ_bb));
+        }
+        ins_vec.push_back(edge.first->getTerminator());
+        return;
+    }
+    ins_vec.push_back(SplitEdge(edge.first, edge.second, dt)->getTerminator());
+}
+
+std::pair<SmallVector<Instruction*,4>,SmallVector<Instruction*,16>>
+LateLowerGCFrame::LocatePushPop(State &S)
+{
+    std::pair<SmallVector<Instruction*,4>,SmallVector<Instruction*,16>> res;
+    auto &push = res.first;
+    auto &pop = res.second;
+    auto global_frame = [&] {
+        push.push_back(&*(++BasicBlock::iterator(ptlsStates)));
+        for (auto &BB: *S.F) {
+            if (auto ret = dyn_cast<ReturnInst>(BB.getTerminator())) {
+                pop.push_back(ret);
+            }
+        }
+    };
+    // These are harder to handle, ignore them for now...
+    if (!S.Allocas.empty() || !S.ReturnsTwice.empty()) {
+        global_frame();
+        return res;
+    }
+    // We don't want to place any push/pop in a loop so first identify the graph of SCC nodes
+    // Each SCC node will always has the same "liveness".
+    struct SCCInfo {
+        std::vector<BasicBlock*> BBs;
+        SetVector<unsigned> Succ{};
+        SetVector<unsigned> Pred{};
```

Comment:
I guess this could also instead just be `SetVector<struct SCCInfo*>`, avoiding the need to number the nodes, if you initialized sccnodes as an eager copy (so you don't need to mutate it later):
```cpp
std::vector<SCCInfo> sccnodes(scc_begin(S.F), scc_end(S.F));
```